api/src/main/java/org/pmiops/workbench/db/dao/DataSetServiceImpl.java

```diff
@@ -171,163 +205,309 @@ public DataSet saveDataSet(
+        getValueSelectsAndJoins(domainValuePairsForCurrentDomain);
+
+    queryBuilder
+        .append(String.join(", ", valuesLinkingPair.getSelects()))
+        .append(" ")
+        .append(valuesLinkingPair.formatJoins());
+
+    final Optional<String> conceptSetSqlInClauseMaybe =
+        buildConceptIdListClause(domain, conceptSetsSelected);
+
+    if (supportsConceptSets(domain)) {
+      final Optional<DomainConceptIdInfo> domainConceptIdInfoMaybe =
+          bigQuerySchemaConfig.cohortTables.values().stream()
+              .filter(config -> domain.toString().equals(config.domain))
+              .map(
+                  tableConfig ->
+                      new DomainConceptIdInfo(
+                          getColumnName(tableConfig, "source"),
+                          getColumnName(tableConfig, "standard")))
+              .findFirst();
+
+      final DomainConceptIdInfo domainConceptIdInfo =
+          domainConceptIdInfoMaybe.orElseThrow(
+              () ->
+                  new ServerErrorException(
+                      String.format(
+                          "Couldn't find source and standard columns for domain: %s",
+                          domain.toString())));
+
+      // This adds the where clauses for cohorts and concept sets.
+      conceptSetSqlInClauseMaybe.ifPresent(
+          clause ->
+              queryBuilder
+                  .append(" WHERE \n(")
+                  .append(domainConceptIdInfo.getStandardConceptIdColumn())
+                  .append(clause)
+                  .append(" OR \n")
+                  .append(domainConceptIdInfo.getSourceConceptIdColumn())
+                  .append(clause)
+                  .append(")"));
+
+      if (!includesAllParticipants) {
+        queryBuilder
+            .append(" \nAND (")
+            .append(personIdQualified)
+            .append(" IN (")
+            .append(cohortQueries)
+            .append("))");
       }
+    } else if (!includesAllParticipants) {
+      queryBuilder
+          .append(" \nWHERE ")
+          .append(personIdQualified)
+          .append(" IN (")
+          .append(cohortQueries)
+          .append(")");
+    }
+
+    final String completeQuery = queryBuilder.toString();
+
+    return buildQueryJobConfiguration(cohortParameters, completeQuery);
+  }
+
+  private void validateConceptSetSelection(Domain domain, List<ConceptSet> conceptSetsSelected) {
+    if (supportsConceptSets(domain)
+        && !conceptSetSelectionIsNonemptyAndEachDomainHasAtLeastOneConcept(conceptSetsSelected)) {
+      throw new BadRequestException("Concept Sets must contain at least one concept");
+    }
+  }
+
+  private Optional<String> buildConceptIdListClause(
+      Domain domain, List<ConceptSet> conceptSetsSelected) {
+    final Optional<String> conceptSetSqlInClauseMaybe;
+    if (supportsConceptSets(domain)) {
+      conceptSetSqlInClauseMaybe =
+          Optional.of(buildConceptIdSqlInClause(domain, conceptSetsSelected));
```

Comment:
Was it really worth breaking out this method? The name of that method is similar but slightly different and it's only ~5 lines of code, and AFAICT not used anywhere else or separately tested or overridden. It's an aesthetic thing, but as a reader I felt more confused than aided by the separation (especially as these methods have no comments, javadocs, or unit tests).

Reply:
Good point, but remember this is merely the opening act for what I'd like to do with this class. I wanted to keep going but had to come up for air in the middle.

The key reason in this case is that I want to highlight that not every (valid) query has a list of concept IDs (thus the Optional). Previously this was implied by concatenating an empty list, but there was lots of confusion around that. I also want to move to a model that encodes this domain knowledge and query parts in a richer way.

Also, the parent method goes from line 367-447. I broke out this one to shrink it slightly, but more importantly, because I understood what it was actually doing and could give it a name. Additionally, lots of these methods are going to be rearranged and put into smaller services and classes, with their own unit tests and things.

All of this is an intermediate state, but I to break down some crazy long functions.

Unit tests are a great idea, as are comments.

Reply:
Added unit tests and consolidated a couple of helper methods.